spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -699,29 +706,83 @@ public DataWriter<InternalRow> createWriter(int partitionId, long taskId, long e
             writeSchema,
             dsSchema,
             targetFileSize,
-            useFanoutWriter);
+            useFanoutWriter,
+            persistLineageFields);
       }
     }
   }
 
-  private static class UnpartitionedDataWriter implements DataWriter<InternalRow> {
+  private abstract static class DataWriterWithLineage implements DataWriter<InternalRow> {
+    private Integer rowIdOrdinal = null;
+    private Integer lastUpdatedOrdinal = null;
+
+    protected InternalRow lineageRow(InternalRow meta) {
+      GenericInternalRow row = new GenericInternalRow(2);
+      row.setNullAt(0);
+      row.setNullAt(1);
+
+      if (meta == null) {
+        return row;
+      }
+
+      // Ordinals are cached
+      if (rowIdOrdinal != null && lastUpdatedOrdinal != null) {
+        setIfNotNull(row, 0, meta, rowIdOrdinal);
+        setIfNotNull(row, 1, meta, lastUpdatedOrdinal);
```

Comment:
I'll ideally need to figure out a shared abstraction for both the MoR DeltaWriter impl and CoW so we can share the logic to derive the metadata values instead of duplicating the logic

Reply:
Ok I think the right thing to do is pass through a `Function<InternalRow, InternalRow> rowLineageExtractor` to the writers and then  we can have a common implementation of that function across CoW/MoR writers.